File: java/matmul.java
class matmul {
	public double[][] matgen(int n) {
		double[][] a = new double[n][n];
		double tmp = 1. / n / n;
		for (int i = 0; i < n; ++i)
			for (int j = 0; j < n; ++j)
				a[i][j] = tmp * (i - j) * (i + j);
		return a;
	}
	public double[][] matmul(double[][] a, double[][] b) {
		int m = a.length, n = a[0].length, p = b[0].length;
		double[][] c = new double[m][p];
		for (int i = 0; i < m; ++i) {
			double[] ci = c[i]; // much faster by hoisting this there
			for (int k = 0; k < n; ++k) {
				double aik = a[i][k];
				double[] bk = b[k];
				for (int j = 0; j < p; ++j)
					ci[j] += aik * bk[j];
			}
		}
		return c;
	}
	public static void main(String[] args) {
		int n = 1000;
		if (args.length >= 1) n = Integer.parseInt(args[0]);
		n = n / 2 * 2;
		matmul m = new matmul();
		double[][] a, b, c;
		a = m.matgen(n);
		b = m.matgen(n);
		c = m.matmul(a, b);
		System.out.println(c[n/2][n/2]);
	}
}

File: src/ocaml/matmul.ml
type mat = float array array

let matgen (n: int) =
  let tmp = 1.0 /. (float_of_int n) /. (float_of_int n) in
  let a = Array.make_matrix n n 0. in
  for i = 0 to n - 1 do
    for j = 0 to n - 1 do
      a.(i).(j) <- tmp *. (float_of_int ((i - j) * (i + j)))
    done
  done;
  a
;;

let matmul (n: int) (a: mat) (b: mat) =
  let c = Array.make_matrix n n 0. in
  for i = 0 to n - 1 do
    let ai = a.(i) in
    let ci = c.(i) in
    for k = 0 to n - 1 do
      let aik = ai.(k) in
      let bk = b.(k) in
      for j = 0 to n - 1 do
        ci.(j) <- ci.(j) +. aik *. bk.(j)
      done
    done
  done;
  c
;; 

let () =
  let n = 1500 in
  let a = matgen n in
  let b = matgen n in
  let c = matmul n a b in
  print_float c.(n/2).(n/2);
  print_newline ()
;;
